### PR TITLE
Implementing Feedback for recomendation algorithm (PS: ignore changes in trie.js. it is being reviewed in a different PR)

### PR DESCRIPTION
**Pull request by Fnuworsu**
# Description  
- What does this PR do?  
  `Implements a hybrid recommendation algorithm combining content-based filtering with collaborative filtering using Matrix Factorization and Gradient Descent`  
- Why is it necessary or valuable?  
  `Enhances content discovery by learning user preferences mathematically, balancing educational content with viral trends and personalization`  
---

# Milestones / Related Work  
- Feature or user story this contributes to: `Implementing Feedback from TC#1`  
---

# Resources and References  
- Tutorials, documentation, or code examples used:  
  `Matrix and Tensor Factorization Techniques applied
to Recommender Systems: a Survey`  [Link](https://d1wqtxts1xzle7.cloudfront.net/78093866/Paper010112-libre.pdf?1641367827=&response-content-disposition=inline%3B+filename%3DMatrix_and_Tensor_Factorization_Techniqu.pdf&Expires=1752705676&Signature=IR3VPBqafew353OgePbkitERJT3Tjg0nIluxbVgLorFVxs~DNqxgavPib~yEFCCvdaoa2ik4q6kIySw0OeglVhAgzWF80jrxR8FrBu1lxqRzXiUd-G~1hJb~nCpH2X936vTd9cR8iHeWqIHTJkVEABQfIEpjvS7Tj5iz5fdwgwSCKfly-ZLQe~xtj4bTbpOmsuyDaPLVmRkhsG-y62bSsSfhhN1RplQm9ZyX~B8xxgmcV4LjdO8JOcOeDh3Ht2G~~u2dRPhtucnI44beuk~aEFzba3dgYY0XCQ2Se52t-fui9aZu8E2J6lXOnsUTOLoDQVxHiRI6dYGtsxdR1KjI1w__&Key-Pair-Id=APKAJLOHF5GGSLRBV4ZA)
- Any copied or adapted code (with attribution):  
  `<Mention if any logic was reused or adapted>`  
- Internal references (Figma, Notion, Slack threads):  
  `<Optional: add internal planning/resource links>`

---
# Algorithm Trade offs
- Recalculation vs. Persistence
     - Current Issue: My algorithm recalculates matrix factorization from scratch on each recommendation request
     - Impact: High CPU usage and latency (5-10 seconds) for each recommendation request
     - Potential solution: Persist trained model to database and load on startup and schedule background updates every few hours
- Memory Usage vs. Cold Starts
     - Current Issue : All user/item matrices are kept in memory with no persistence
     - Impact: Server restarts lose all learned preferences, forcing "cold start" each time
     - Potential solution: Implement model serialization to database and deserialize on startup
- Batch Processing vs. Real-time Updates
     - Current Issue: `updateUserInteraction()` retrains the entire model after each user interaction 
     - Impact: High CPU usage after user actions (likes/comments)
     - Potential Solution**: Queue interactions and process in batches and use incremental updates for matrix factorization
---

# Test Plan  
- How was this tested?  
  `Manual testing with sample user-item interaction data and synthetic engagement metrics`  

## Edge Cases Tested  
*   `Cold start: New users with no interaction history`
*   `Sparse data: Users with minimal interactions`

---

# Additional Notes   
- Reviewer requests or feedback needed: `Feedback on gradient descent parameters and convergence threshold values`

### CODE REVIEW

File: backend/routes/searchRoutes/searchRoutes.js [hunk 1]
```diff
@@ -6,9 +6,9 @@ const router = express.Router()
 const prisma = new PrismaClient()
 
 // Search across all content types
-router.get('/api/search', isAuthenticated, async (req, res) => {
+router.post('/api/search',isAuthenticated, async (req, res) => {
```
Comment (dorcheng) [inline backend/routes/searchRoutes/searchRoutes.js:9]: How come you're changing "get" to "post"? The standard practice is to use "get" when you're trying to fetch something from server and to use "post" when you're trying to submit data to the server to add/modify/delete something

File: backend/typeAhead/trie.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            }
+        }
+    }
+
+    dfs(obj)
+    return result
+}
+// manual test
+const words = extractAllStrings(data)
+console.log(autoComplete(words, "build"))
```
Comment (dorcheng) [inline backend/typeAhead/trie.js:85]: This looks like it was just used for testing so can be removed in the committed code :)

File: backend/routes/searchRoutes/searchRoutes.js (hunk 1 above)
Comment (Fnuworsu) [reply backend/routes/searchRoutes/searchRoutes.js:9]: @dorcheng I was using req.body and during api testing, the api tester I was using did not allow using request body for get requests

File: backend/typeAhead/trie.js (hunk 2 above)
Comment (Fnuworsu) [reply backend/typeAhead/trie.js:85]: @dorcheng , Noted!
